CMaxTable.lua

```diff
@@ -19,8 +19,7 @@ end
 
 function CMaxTable:updateGradInput(input, gradOutput)
    for i=1,#input do
-      self.gradInput[i] = torch.Tensor()
-      self.gradInput[i]:resizeAs(input[i]):fill(0.0)
+      self.gradInput[i] = input[i]:clone():fill(0.0)
```

Comment:
Instead of cloning and then zeroing the tensor, it is more efficient to creat an empty tensor using `input[i].new()`.
Also, you are creating new tensors at every backward pass, which is a blocking operation and currently slows down GPU computation a lot.
Doing something like
```lua
self.gradInput [i] = self.gradInput [i] or input [i].new ()
```
will avoid repeated memory allocations.